### ISSUE TITLE
Disabling snippets included from packages

### ISSUE THREAD

**Issue by kyeweedon** [atom/snippets#55]
For example, the `language-javascript` package includes 24 default snippets, many of which I have developed alternatives for which better match my coding style.

How can I block this package from polluting the snippets selection I've defined elsewhere?
# help wanted

**Comment by jonathandelgado** [comment]
+1 - This would really be a fantastic feature.

**Comment by Fank** [comment]
:+1:

**Comment by pdostal** [comment]
:+1:

**Comment by wyand** [comment]
:+1:

Perhaps this could be implemented by an addition to the `snippets.cson` format?

```coffee
'.source.js':
  'setInterval function':
    disabled: true
```

Possible issues on that might be that its inconvenient or unstable to disable by name instead of prefix, and more so the order of operations. Using `disabled` in a package's snippets, instead of in `~/.atom/snippets.cson` or `~/.atom/snippets` could be difficult to manage consistently.

**Comment by atom-project** [comment]
:thumbsup:

**Comment by varemenos** [comment]
:+1: this would really help declutter snippet lists with useless stuff that some of us don't want.

**Comment by evanwarner** [comment]
+1

**Comment by Janghou** [comment]
More Unix-wise:
Why not rename the snippets directory to `snippets_disabled`  or file to `cson_disabled`. It will skip indexing. That's how I do it now manually. 

Just add a setting to rename and tweak the update process to take it in account.

PRO: you have to transparent ways of handling it, manually or through the GUI. 

That's hacking of the good old days.

**Comment by jprichardson** [comment]
Would love this. Just switched over to Atom from Sublime and some of the Atom snippets are super annoying that I'd like to disable but keep the grammar enabled for the sake of syntax highlighting.

**Comment by benatkin** [comment]
:+1:

**Comment by xicombd** [comment]
+1

**Comment by zyklus** [comment]
+1

**Comment by michaeljacobdavis** [comment]
:+1:

**Comment by oguzbilgic** [comment]
+1

**Comment by oraculus** [comment]
:+1:

**Comment by 50Wliu** [comment]
This could probably use the same logic as the recent Keybindings PR.

**Comment by mangecoeur** [comment]
+1 - the snippets can be super annoying, for instance if you are editing plain text. The Autocomplete plugin allows you to disable lower priority plugins using a higher priority one which means you can at least muck about with priorities to get the behaviour you want, but the snippets have no such option which can get super annoying.

**Comment by rearadmiral** [comment]
:+1:

**Comment by lesti** [comment]
+1

**Comment by fumieval** [comment]
+1

**Comment by DaleJefferson** [comment]
This is such a basic feature, maybe we could split the grammar from the snippets, they are really two separate things. I should be able to get the Javascript grammar without the snippets.

**Comment by vemv** [comment]
I wonder where the $200M GitHub raised are going... incredible this feature is still missing.

**Comment by brightbits** [comment]
+1 I'd love to turn off snippets for certain grammars. In ruby when I type `:id` and press return it turns it in to `:# IDEA:` which is so total not what I want to happen. Same thing with many other snippets being really annoying. If I have an idea I'll write a story for it, not put it in a random comment somewhere :/

**Comment by ystkme** [comment]
+1  
I really want this feature

**Comment by jpillora** [comment]
+1

**Comment by bohdan-chechin** [comment]
+1

**Comment by RangerMauve** [comment]
+1 Relating to Flow snippets in https://github.com/Icehunter/language-jsx/issues/4

**Comment by bsr203** [comment]
+1

**Comment by jeremyfa** [comment]
+1

**Comment by mayowa** [comment]
+1

**Comment by mehwww** [comment]
+1

**Comment by bakape** [comment]
+1

**Comment by dimitri-xyz** [comment]
+1
It would be great to be able to disable snippets.

**Comment by jameskopacz** [comment]
:+1:

**Comment by alexfedoseev** [comment]
:+1:

**Comment by kyaido** [comment]
+1

**Comment by RedFranck** [comment]
+1

**Comment by andgudovich** [comment]
+1

**Comment by theorygeek** [comment]
+1

**Comment by TheGreatPanda** [comment]
+1 we need this

**Comment by matteddy** [comment]
+1

**Comment by dralletje** [comment]
Instead of +1'in it, I thought I could just give it a shot :grin: #196

**Comment by Kaijun** [comment]
+1!!!!!!!!

**Comment by Kaijun** [comment]
@dralletje this would be cool!!!! hope get it in the next release!

**Comment by ttimasdf** [comment]
Supposing I'm going to disable a python snippet from official `language-python` named `"self"` with prefix `"."`, as it conflicts with `autocomplete-python` trigger. Just add the following to the `snippets.cson`
```cson
'.source.python':
  'whatever the name':
    'prefix': '.'
    disabled: true
```

Inspired by [his](#issuecomment-102840442) comment 😉

**Comment by mehwww** [comment]
+1

**Comment by JeremyIglehart** [comment]
+1

**Comment by ripter** [comment]
Some of the current snippets like require only work if you have bad habits. Example: if you want `var foo = require('./foo.js');`, the snippet only activates when you start typing require, which will result in `var foo = var module = require('module');`.

The only way this snippet would be correct is if you don't type the `var foo =` part and start with `require`.  Since I have good habits and type what i want instead of relying on a snippet to type the correct code for me, the snippet is just annoying and never works properly.

**Comment by dralletje** [comment]
@ripter 

That is not about 'bad habits' but about a different way of using the snippets. You are using it like autocomplete, where snippets are usually more intended as actual... snippets. The 'coincidence' that the snippet triggered by the word `require` does also produce that some word in it's output is merely coming from the fact that the name `require` just fits so well for the snippets and the code the same.

If you came here just to say that you want the feature to disable snippets, you could just have said that. Next time you call something "bad habits", please reconsider.

**Comment by VladShcherbin** [comment]
Jesus, even this feature is not present.. :(

**Comment by joeroback** [comment]
:+1: but the reality for me is, i have autocomplete for all the languages i use with atom, so i find the snippets annoying at best, especially being mixed in with autocomplete, so i just disable the autocomplete-snippets package entirely and solves 99.9% of issues. it's a big hammer, but makes my editing experience at this point an order of magnitude better IMHO

**Comment by reqw** [comment]
@50Wliu will it be ever added?

**Comment by 50Wliu** [comment]
No one's created a full PR for it yet.

**Comment by danielbayley** [comment]
There should be an automatic checkbox in settings for every package that includes snippets, using exactly the same system as keybindings, and I would stretch that to include menu/context menu items as well.

**Comment by 50Wliu** [comment]
@danielbayley I agree with that (not necessarily the menu part though), and it's what I [suggested](https://github.com/atom/snippets/issues/55#issuecomment-136214008) after that PR got merged.  Well thought-out PRs would be appreciated.

**Comment by paulmsmith** [comment]
What's going on with this? @50Wliu? I notice there is a PR for this but has it not been merged yet? Is there a manual way I can disable snippets for individual languages? Thanks for all the hard work.

**Comment by 50Wliu** [comment]
If you're referencing #196 - that PR is missing specs.
If you're referencing #216 - that PR is missing a per-package toggle.

I am not aware of any manual way of disabling snippets.

**Comment by Hacalox** [comment]
Any news related to this feature? Is there any way to prioritize my snippets over the default snippets? 
Thank you!

**Comment by Hum4n01d** [comment]
Yes PLEASE

**Comment by PierBover** [comment]
Is anyone working on this? Is there still interest?

**Comment by dralletje** [comment]
Hey @PierBover , as far as I know my PR is still under need-review, after I added specs to it :-)
I understand the contributors have their hands full with lots of stuff and open source work is not high priority, still it's sad to see so many people want the feature, hahaha

https://github.com/atom/snippets/pull/196

**Comment by PierBover** [comment]
Yes @dralletje they are pretty busy.

Maybe pinging @50Wliu and @lee-dohm will help :)

**Comment by dzift** [comment]
[joeroback's comment](https://github.com/atom/snippets/issues/55#issuecomment-241042266) basically solved my problem. 

Disabling 'autocomplete-snippets' package removes snippet suggestions as you type, yet doesn't disable them.

**Comment by jeanfredrik** [comment]
👍

**Comment by vczhan** [comment]
👍

**Comment by L1Q** [comment]
+1

**Comment by dzift** [comment]
As a workaround I disabled the autocomplete-snippets package.

Also in the autocomplete-plus package, setting the 'Keymap for Confirming a Suggestion' to only 'Enter' made my life easier.

Hope the whole snippet/autocomplete business gets reworked asap.

**Comment by ffxsam** [comment]
@dzift Thanks for that. For me, setting it to "tab" made it tolerable for me. Whenever I'd type "id" and hit enter, it would autocomplete to "// IDEA".

**Comment by ThomasChaf** [comment]
In my snippets.cson I did:

```
'.source.XX':
  'workaround id':
    'prefix': 'id'
    'body': 'id'
```

At least when I type "id" and then press enter/tab it doesn't automatically include some undesired text. I'm not really happy about this but I'm just waiting one of the PR to be included in atom/snippets. 

I did the same with 'review'.

**Comment by chasevida** [comment]
@wyand little counter snippet worked for me in Atom `1.18.x`. I was able to obliterate this annoying default snippet which was constantly interfering with day to day work.

```
'.source.js':
  'xxx - please no!':
    'prefix': 'xxx'
    disabled: true
```

**Comment by benpptung** [comment]
Today, I've added more useful packages, but also got more unwanted crazy snippets...I love snippet, but also hate snippet...

If it's difficult to find where snippets coming from and disable them, how about add the following feature:

## disable all package snippets except personal snippets ?

I think it's really easy to add snippet using `snippet.cson`, if some snippets are good, we can add them ourselves.

**Comment by kunokdev** [comment]
any progress on this so far?

**Comment by dralletje** [comment]
@kunokdev AS A MATTER OF FACT!!!!! #242

**Comment by vimuel** [comment]
Is this coming any time soon?

**Comment by 50Wliu** [comment]
We are waiting for changes to be made in #242.

**Comment by brianpurkiss** [comment]
+1 to this. 

Whenever I type in "margin-t" (in CSS) it auto completes to the //todo auto complete instead of "margin-top" - similar annoying auto completes happen for "color" and other commonly used CSS styles and it auto completes to things I don't ever use. 

This makes auto-complete a hindrance instead of a help for much of the CSS I commonly write. 

If I could just get the "temp" and "todo" and "combak" auto completes to be removed or the last priority instead of the first priority I would absolutely LOVE it. I've been tempted to turn off auto complete many times.

**Comment by leelandmiller** [comment]
+9000

I'm so close to switching editors because of how annoying some snippets are to me. They end up slowing me down by throwing random, unwanted crap into my code.

**Comment by kunokdev** [comment]
So yeah, #242 is kinda canceled.

**Comment by leelandmiller** [comment]
@brianpurkiss 
I was having problems with those exact snippets, and finally figured out a way around it.
I wanted to keep the syntax highlighting for my todo's, temp's, and idea's but the snippets were just too much.

So here's what I did (and this will work for any language package):
1.) Went to the repo `/atom/language-todo`
2.) Clone `language-todo` into `.atom/packages`
3.) Open `.atom/packages/language-todo/snippets/todo.cson`
4.) Delete the undesired snippets.

**Comment by Beefster09** [comment]
What's holding this back? This seems like an obvious feature to implement and I think it ought to be higher priority.

A feature as essential as this shouldn't take 3 years to get to.

**Comment by lee-dohm** [comment]
Thanks for your feedback @Beefster09.

Open source software depends on contributions from the community in order for it to reach its maximum potential. We can understand that some people will not agree with our choices in how to prioritize our work. One choice people have is to contribute their own work to make something happen, whether that is in the form of a pull request, a design of a solution, creating an Atom package, or pulling together a team to create the solution. Working together to make awesome stuff is what open source is all about.

Demands for status, declarations of how long something has been open, requiring us to justify our prioritization, or other antagonistic requests for response to an issue or PR are all considered personal attacks under our [Code of Conduct](https://github.com/atom/atom/blob/master/CODE_OF_CONDUCT.md). Let's keep things classy instead.

**Comment by T-RegEx** [comment]
I'm desperately waiting for it! 
Or will this feature only be implemented after VCode has made it possible to explicitly change the font size for the file tree? ... ;-)

**Comment by savetheclocktower** [comment]
FYI: This enhancement is on track to be included in the 1.33 release. Thanks to everyone who weighed in so that we could close this ticket before it became old enough to attend kindergarten.
